### DISCUSSION TITLE
SYSEX:Message:-1 with 'F0...LS MS F7': -1 = [7F FF], instead of [F7 F7]

### DISCUSSION THREAD

**Discussion by unityconstruct** [RomanKubiak/ctrlr#628] [Q&A] [answered]
- Have a volume control that is [-96 -> 0 -> +10 ] ( -96,10 )
- EMU uses 2's compliment for negatives
- my function formula nibblizes -1 as expected for [LS MB] 7bit protocol, but the built-in sysex seems to go off the rails
- code shows that these should be on the money, but the [MS] is yielding 255, not 127
  - that that translates to LS =7bit  &  MS=8bit... 
  - making MS a LIAR.. but I'm sure I'm doing something stupid.

# Anyone able to slap me around & back on track with what I'm missing here?
- trying to avoid creating another function for this as I have oodles of mods running negatives.
- tested on `Prodatum` to confirm my understanding of the device requirements/sysex spec & these messages correctly traverse the 03 -> -03 boundary:
### Correct messages from Prodatum
```
F0 18 0F 00 55 01 02 02 0B 03 00 F7
F0 18 0F 00 55 01 02 02 0B 02 00 F7
F0 18 0F 00 55 01 02 02 0B 01 00 F7
F0 18 0F 00 55 01 02 02 0B 00 00 F7
F0 18 0F 00 55 01 02 02 0B 7F 7F F7
F0 18 0F 00 55 01 02 02 0B 7E 7F F7
F0 18 0F 00 55 01 02 02 0B 7D 7F F7
F0 18 0F 00 55 01 02 02 0B 7C 7F F7
```
## 
| token | type |
|-- | -- |
| LS | LSB7bitValue |
| MS | MSB7bitValue |

![ctrlr-slider-neg-values](https://github.com/RomanKubiak/ctrlr/assets/40268430/423be717-9377-45d2-85c8-d76150000baf)

### Sysex tokens

`*` = one or more ANY characters

| token | type |
|-- | -- |
| xx | ByteValue |
| yy | ByteChannel |
| LS | LSB7bitValue |
| MS | MSB7bitValue |
| ii | Ignore |
| ls | LSB4bitValue |
| ms | MSB4bitValue |
| r1 | RolandSplitByte1 |
| r2 | RolandSplitByte2 |
| r3 | RolandSplitByte3 |
| r4 | RolandSplitByte4 |
| tp | CurrentProgram |
| tb | CurrentBank |
| *y | ByteChannel4Bit |
| z* | ChecksumRolandJP8080 |
| w* | ChecksumWaldorfRackAttack |
| u* | LUAToken |
| v* | FormulaToken |
| k*, p*, n*, o* | GlobalVariable |

**Comment by damiensellier** [comment]
Many synths use 0-255 split on 2 bytes for negative/positive range values.
And it seems your bytes are nibblized by pair where 0xff is actually 0x0f 0x0f.

maybe you should take a look at those methods if it relates : 
https://github.com/RomanKubiak/ctrlr/discussions/533
https://github.com/RomanKubiak/ctrlr/discussions/250#discussioncomment-3865023
https://github.com/RomanKubiak/ctrlr/discussions/515

Let us know if you get something from the discussions/posts above ;)

**Reply by unityconstruct** [reply]
> "Negativ values again :) "  <<<---- 
haha... after wading through the topic a gorrillion times in posts and docs... & always completely understand it....

.. until I don't - normally with that obligatory week commencing only hours after an epiphany that rips one way from the project long enough to lose fluency

**Comment by dnaldoog** [comment]
Hi UC,

I've encountered this with the Korg panels I have worked on:

They (e.g `7f 7f`) are sysex representations of 8 bit signed integers.

Here is the code I came up with a while back. I stopped thinking about it once I got it working, so maybe there's a better way? For example splitting hex strings with [bit.tohex](https://bitop.luajit.org/api.html) but I figure my algorithm is faster.

But I see @damiensellier has come up with some old solutions anyway.

<pre>
myMethod = function(--[[ CtrlrModulator --]] mod --[[ number --]], value --[[ number --]], source)
    local MSB, LSB = signedShort(value)
    local sys = MemoryBlock(string.format("F0 18 0F 00 55 01 02 02 0B %.2X %.2X F7", MSB, LSB))
   panel:sendMidiMessageNow(CtrlrMidiMessage(sys:toHexString(1)))
end

signedShort = function(input)
    local l, m = bit.band(127, input), input < 0 and 127 or 0
    --return l,m -- korg order of returning LSB.MSB
    return m, l -- maybe other manufacturers go this way?
end
</pre>

![Untitled](https://github.com/RomanKubiak/ctrlr/assets/76831683/ff2d9321-4b56-40f3-a434-55fe612f9962)


[signed Short 2 byte_1_0_Hell-O-Kitty.zip](https://github.com/RomanKubiak/ctrlr/files/14785009/signed.Short.2.byte_1_0_Hell-O-Kitty.zip)

**Comment by dnaldoog** [comment]
If you want to avoid creating another function, this is also pretty terse:

<pre>
myMethod = function(--[[ CtrlrModulator --]] mod --[[ number --]], value --[[ number --]], source)
    local LSB = bit.band(127,value)
    local MSB = bit.band(128,value) == 128 and 127 or 0 
    local sys = MemoryBlock(string.format("F0 18 0F 00 55 01 02 02 0B %.2X %.2X F7", MSB, LSB))
    panel:sendMidiMessageNow(CtrlrMidiMessage(sys:toHexString(1)))
end
</pre>

**Comment by dnaldoog** [comment]
And here is how you would do it with `bit.tohex()`:

<pre>
myMethod = function(--[[ CtrlrModulator --]] mod --[[ number --]], value --[[ number --]], source)
local LSB = bit.tohex(bit.band(127,value),-2)
local MSB = bit.tohex(bit.band(128,value)/128*127,-2)
local sys = MemoryBlock(string.format("F0 18 0F 00 55 01 02 02 0B %s %s F7", MSB, LSB))
panel:getLabelComponent("output"):setText(sys:toHexString(1):upper())
panel:sendMidiMessageNow(CtrlrMidiMessage(sys:toHexString(1)))
end
</pre>

**Comment by unityconstruct** [comment]
THANKS for info! Here's E-MU Sysex Spec.. the issue might not be that field in Ctrlr, but just my bad assumption.. In hindsight.. that field works awfully well for CCs... which are 127.... so that would imply they don't work too well with 256. Looks like I'm trying to avoid the inevitable.
- All my params are suffixed with a param id so should be able to just have them all pile into one function & let God sort them out there.

THANKS!

![image](https://github.com/RomanKubiak/ctrlr/assets/40268430/572ae9c9-e53c-4ef9-80d3-2d4648f19e60)


- BTW Getting close to `Mark I` of this panel... just a few more functional proof-of-concept UI pieces... your mentoring has been immensely time-saving & helpful.. particularly, `the elder posts` on ctrlr.org. Reverse-engineering Ctrlr with no reference would have been brutal. The value of those PDF guides can't be overstated. So sincere appreciation for all dat. It would figure I went for the most complex architecture in the house to get my feet wet 😫 - of course considering [2x24" monitors] VS [2x24 chars] + in-DAW automation, the scars are well worth the effort.
## The 8 week crash course so far: admittedly not as slick as the Ctrlr OG's here🤫
![2024 03 27 212013 cap](https://github.com/RomanKubiak/ctrlr/assets/40268430/6fa437d3-ffc1-4ab6-85e2-3b8414140ea2)

![2024 03 27 212208 cap](https://github.com/RomanKubiak/ctrlr/assets/40268430/99f516ed-5406-4a33-afc2-927f6576544f)

**Comment by dnaldoog** [comment]
How about this panel? This is for 14 bit -ve/+ve (-8192 ~~ + 8191)

This code follows the manual screenshot you posted as closely as possible, but I wonder if those negative values are correct for large negative values?

<pre>
myMethod = function(--[[ CtrlrModulator --]] mod --[[ number --]], value --[[ number --]], source)
    local MSB, LSB = 0, 0
    if value < 0 then
        value = value + 16384
        MSB = value / 128
        LSB = value % 128
    else
        MSB, _ = math.modf(value / 128)
        LSB = bit.band(127, value)
    end
    local sys = MemoryBlock(string.format("F0 18 0F 00 55 01 02 02 0B %.2x %.2x F7", MSB, LSB))
end
</pre>
----
[signed 14bit panel_1_0.zip](https://github.com/RomanKubiak/ctrlr/files/14789071/signed.14bit.panel_1_0.zip)

**Reply by unityconstruct** [reply]
hey brother! too funny... I recall trying to get `bit` on my lua install like on Day#3 of my Lua career and couldn't figure it out...  ( likely due to all the bemoaning over being forced "to learn such a crappy language that has nothing in it to make life easy"😂  ) ...
..now come to find out CTRLR already has it embedded. Well at least I know how to do it manually for the next job interview 'gotcha' quiz.

just now I tried again & built/used it in 10m without issue. Never ceases to amaze me how stuff that was once impossible, in the end is trivial.

checking out the panel now sir.. thx!

**Reply by dnaldoog** [reply]
So does that mean this latest code is correct? 🫤

**Comment by dnaldoog** [comment]
I think this is even more efficient code:

<pre>
myMethod = function(--[[ CtrlrModulator --]] mod --[[ number --]], value --[[ number --]], source)
        local value = value + 16384
        local MSB = value > 16383 and (value/128)-128 or value / 128
        local LSB = value % 128
    local sys = MemoryBlock(string.format("F0 18 0F 00 55 01 02 02 0B %.2x %.2x F7", MSB, LSB))
end
</pre>

[signed 14bit panel_2_0.zip](https://github.com/RomanKubiak/ctrlr/files/14798580/signed.14bit.panel_2_0.zip)


![14bitnegative](https://github.com/RomanKubiak/ctrlr/assets/76831683/6dfd94df-d3ae-448d-9291-f26fe12061ac)

**Comment by dnaldoog** [comment]
This panel shows how to also convert an incoming Midi message with nibbleized 14 bit (2 byte values) - This code would assign a value to a modulator from incoming MIDI (in the panel I am just using the contents of the `uiLabel` as a dummy incoming message).

The top `uiSlider` shows how to generate the 2 bytes and the 'convert' button shows how you would convert 2 bytes to a +/- value updating a modulator `uiSlider` etc.

<pre>
midiReceived = function(--[[ CtrlrMidiMessage --]] midi)
    local rawValue = (midi:getData():getByte(9) * 128) + midi:getData():getByte(10)
    local conversion = rawValue > 8191 and rawValue - 16384 or rawValue
    panel:getModulatorByName("some target"):setModulatorValue(conversion,false,false,false)
end
</pre>

![14bitnegativeversion3](https://github.com/RomanKubiak/ctrlr/assets/76831683/7c2f35bd-34de-4c27-bad1-49d1dce2dd0e)

[signed 14bit panel_3_0.zip](https://github.com/RomanKubiak/ctrlr/files/14799082/signed.14bit.panel_3_0.zip)

**Comment by dnaldoog** [comment]
Still going!

Here's a one liner! Somebody stop me!

<pre>
myMethod = function(--[[ CtrlrModulator --]] mod --[[ number --]], value --[[ number --]], source)
local MSB,LSB=split7bit(value)
    local sys = MemoryBlock(string.format("F0 18 0F 00 55 01 02 02 0B %.2x %.2x F7", MSB, LSB))
     panel:sendMidiMessageNow(CtrlrMidiMessage(sys:toHexString(1)))
end


split7bit=function(n)
return bit.rshift(bit.band(n,0x3f80),7),bit.band(n,127)
end
</pre>

----
[signed 14bit panel_4_0.zip](https://github.com/RomanKubiak/ctrlr/files/14812048/signed.14bit.panel_4_0.zip)

**Reply by Ted-juh** [reply]
No, go ahead, you're on a roll.. 😜

Bet you can't do it with no lines 😉

**Comment by dnaldoog** [comment, marked as answer]
Finally, *possibly finally*, I think the following may be the best code, because it is simpler:

<pre>
split7bit = function(n)
    local bi = BigInteger()
    bi:setBitRangeAsInt(0, 14, n)
    return bi:getBitRangeAsInt(7, 7),bi:getBitRangeAsInt(0, 7)
end
</pre>

The reason why I never tried `BigInteger` is because initializing (which I usually do) with `local bi=BigInteger(n)` gives a different result for negatives, I think because of this note on the [JUCE](https://docs.juce.com/master/classBigInteger.html) site: ↴
#### Negative values are possible, but the value isn't stored as 2s-complement, so be careful if you use negative values and look at the values of individual bits.
But for some reason initialising with `setBitRangeAsInt()` seems to work?? Go figure. 

You can see this with `bi=BigInteger(-1)` and `bi2=BigInteger() bi2:setBitRangeAsInt(0,14,-1)`
The first one `bi:toString(2,0)` prints `-1` whereas `bi2:toString(2,0)` prints `11111111111111`

[signed 14bit panel_5_0.zip](https://github.com/RomanKubiak/ctrlr/files/14811360/signed.14bit.panel_5_0.zip)

**Reply by unityconstruct** [reply]
@dnaldoog  ...Apologies for delayed emoji sir... was in crazy-ctrlr-crashtown implementing PopUpMenu<--> uiLabel for an ungodly number of params... and survived that only to find its a solution for 5-6 other things so had to go from 'just make it work' to 'make it work correctly... consistently'.

- Doubling back on your mentoring now.. THANKS for the details!!! Dissection in progress.

> [dnaldoog](https://github.com/dnaldoog)
> [2 weeks ago](https://github.com/RomanKubiak/ctrlr/discussions/628#discussioncomment-8946298)
> So does that mean this latest code is correct? 🫤

- yessir indeed - off the cuff... (for E-MU at least) their 14bit LSB/MSB logic has max 16K values with negatives effectively getting summed with 16K before nibblitastic ops
```lua
tonumber("1111111",2) -- 7bit  > int
127 -- (128 possible values in 7bit byte)
128*128 = 16384 --(in 2 7bit bytes)
```
- this perks my interest!.
>```lua
>split7bit = function(n)
>    local bi = BigInteger()
>    bi:setBitRangeAsInt(0, 14, n)
>    return bi:getBitRangeAsInt(7, 7),bi:getBitRangeAsInt(0, 7)
>end
>```
- it is obtuse on the eyes, but getting stuff like this down to terse statements can really help clean up my code.
- I don't have `bit` available outside Ctrlr, but I think that first call returns `self` to facilitate chaining
- its a  ***`bit`*** ( get it :laughing: ) redundant, but we might have a one-liner for the anonymous-functionally-challenged like myself :smile_cat: 
```lua
local msb, lsb = BigInteger():setBitRangeAsInt(0, 14, n):getBitRangeAsInt(7, 7), BigInteger():setBitRangeAsInt(0, 14, n):getBitRangeAsInt(0, 7)
```
